keps/sig-storage/20190709-csi-snapshot.md

```diff
@@ -89,28 +92,150 @@ The rest of the document includes required information missing from the original
+* VolumeSnapshotSpec has a new required Source field. Source may be either a PersistentVolumeClaimName (if dynamically provisioning a snapshot) or VolumeSnapshotContentName (if pre-provisioning a snapshot).
+* VolumeSnapshotContentSpec also has a new required Source field. This Source may be either a VolumeHandle (if dynamically provisioning a snapshot) or a SnapshotHandle (if pre-provisioning volume snapshots).
+* VolumeSnapshotStatus now contains a BoundVolumeSnapshotContentName to indicate the VolumeSnapshot object is bound to a VolumeSnapshotContent.
+* VolumeSnapshotContentnow contains a Status to indicate the current state of the content. It has a field SnapshotHandle to indicate that the VolumeSnapshotContent represents a snapshot on the storage system.
+
+The beta VolumeSnapshot API object is as follows:
+
+```
+type VolumeSnapshot struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Spec VolumeSnapshotSpec
+        Status *VolumeSnapshotStatus
+}
+```
+
+```
+type VolumeSnapshotSpec struct {
+	Source VolumeSnapshotSource
+	VolumeSnapshotClassName *string
+}
+// Exactly one of its members MUST be specified
+type VolumeSnapshotSource struct {
+	// +optional
+	PersistentVolumeClaimName *string
+	// +optional
+	VolumeSnapshotContentName *string
+}
+```
+
+```
+type VolumeSnapshotStatus struct {
+	BoundVolumeSnapshotContentName *string
+	CreationTime *metav1.Time
+	ReadyToUse *bool
+	RestoreSize *resource.Quantity
+	Error *VolumeSnapshotError
+}
+```
+
+The beta VolumeSnapshotContent API object is as follows:
+
+```
+type VolumeSnapshotContent struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Spec VolumeSnapshotContentSpec
+        Status *VolumeSnapshotContentStatus
+}
+```
+
+```
+type VolumeSnapshotContentSpec struct {
+         VolumeSnapshotRef core_v1.ObjectReference
+         Source VolumeSnapshotContentSource
+         DeletionPolicy DeletionPolicy
+         Driver string
+         VolumeSnapshotClassName *string
+}
+```
+
+```
+type VolumeSnapshotContentSource struct {
+	// +optional
+	VolumeHandle *string
+	// +optional
+	SnapshotHandle *string
+}
+```
+
+```
+type VolumeSnapshotContentStatus struct {
+  CreationTime *int64
+  ReadyToUse *bool
+  RestoreSize *int64
+  Error *VolumeSnapshotError
+  SnapshotHandle *string
+}
+```
+
+The beta Kubernetes VolumeSnapshotClass API object is as follows:
+
+```
+type VolumeSnapshotClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Driver string
+        Parameters map[string]string
+        DeletionPolicy DeletionPolicy
+}
+```
+
+### Controller Split
+
+When Volume Snapshot is promoted to Beta in Kubernetes 1.17, the CSI external-snapshotter sidecar controller is split into two controllers: a snapshot-controller and a CSI external-snapshotter sidecar.
```

Comment:
"Along with VolumeSnapshot being promoted to Beta in Kubernetes 1.17, the CSI external-snapshotter sidecar controller has been split into two controllers:"